### PR TITLE
Lock down neighbourhoods

### LINKED ISSUE #2053: [Bug]: Citizen users with neighbourhood permissions can edit any neighbourhood
## Description

Users with neighbourhood_admin roles can edit any neighbourhood. This feels like something we should lock down - the neighbourhoods are all dependent on each other in ways that may not be clear to a user and changing a field could have wide reaching consequences that they do not understand, and impact on other users' inclusion in sites.

It also looks like they can edit a name of a neighbourhood & the timestamps but these changes do not save - but this is a redundant concern because I think we should essentially just remove the edit view from them completely.

## Acceptance criteria

- [ ] citizen user with a neighbourhood can visit neighbourhoods index
- [ ] citizen user with a neighbourhood can view their neighbourhood
- [ ] citizen user with a neighbourhood can view other neighbourhoods, but not see their details
-----
- [ ] citizen user with a neighbourhood cannot edit their neighbourhood
- [ ] citizen user with a neighbourhood cannot destroy neighbourhoods
-----
- [ ] root user with a neighbourhood can do all of the above, including for neighbourhoods they do not own
- [ ] root user without a neighbourhood can do all of the above

### Plan

Write tests for all of the above AC:
- [x] citizen user with a neighbourhood can visit neighbourhoods index
- [x] citizen user with a neighbourhood can view their neighbourhood
- [x] citizen user with a neighbourhood cannot edit their neighbourhood
- [x] citizen user with a neighbourhood cannot view other neighbourhoods
- [x] citizen user with a neighbourhood cannot destroy neighbourhoods
- [x] root user with a neighbourhood can do all of the above, including for neighbourhoods they do not own
- [x] root user without a neighbourhood can do all of the above
- Remove links to pages that should not be visible
- add policies to controllers for pages that should not be visitable

### PR DESCRIPTION
**Pull request by r-ferrier**
Fixes #2053

Neighbourhood admins rejoice! Your permissions have been "enhanced". You can now no longer edit neighbourhoods, or view details about neighbourhoods you have not been assigned to. 

You can still navigate the big neighbourhoods tree, and view very basic details about each neighbourhood within it on their show page. If you admin for any of those neighbourhoods, you can see slightly more details on this page. 

I also fixed the partner scope which was preventing neighbourhood admins with partners in their neighbourhood from being able to edit partners, ie. assign them admins after creation. 

I think we might want to investigate locking down other partner details as a separate ticket to prevent neighbourhood admins from being able to make edits to partners beyond the assignation of admin duties.

To test this I made sure all the acceptance criteria were written down in the test suite in the form of tests, and tried to make sure that as a neighbourhood_admin with very few other permissions I can still take all the actions I need to. I definitely recommend whoever reviews this do the same as there will no doubt be some missed edge cases!

### CODE REVIEW

File: app/datatables/neighbourhood_datatable.rb [hunk 1]
```diff
@@ -18,7 +18,7 @@ def view_columns
   def data
     records.map do |record|
       {
-        id: link_to(record.id, edit_admin_neighbourhood_path(record)),
+        id: options[:current_user].can_view_neighbourhood_by_id?(record.id) ? link_to(record.id, admin_neighbourhood_path(record)) : record.id,
```
[the remark this answers is not in the document]
Comment (r-ferrier) [reply app/datatables/neighbourhood_datatable.rb:21]: I don't like to reassign a variable unless asbolutely necessary but I take your point about the messiness, so will just move this logic into a multi line if/else/end statement above the hash

**Comment by r-ferrier** [comment]
I've made the requested changes, and I * think * the service area assignment for partners is working as it should - you can only add service areas from within your neighbourhood, unless you admin for a partner, in which case you can see everything. Let me know if you need any more changes!

**Comment by aaaaargZombies** [comment]
> I've made the requested changes, and I * think * the service area assignment for partners is working as it should - you can only add service areas from within your neighbourhood, unless you admin for a partner, in which case you can see everything. Let me know if you need any more changes!

Maybe this is desired behavior? the only think I notice is that a `neighbourhood_admin` can only see partners that are in their neighbourhood, so being able to add a service area for a partner that already belongs to the neighbourhood is a redundant.

It does prevent a `neighbourhood_admin` from adding lots of partners to someone else's neighbourhood that they have not been given admin privileges for so I suppose it is an improvement for the rival `neighbourhood_admin`

**Comment by r-ferrier** [comment]
> > I've made the requested changes, and I * think * the service area assignment for partners is working as it should - you can only add service areas from within your neighbourhood, unless you admin for a partner, in which case you can see everything. Let me know if you need any more changes!
> 
> Maybe this is desired behavior? the only thing I notice is that a `neighbourhood_admin` can only see partners that are in their neighbourhood, so being able to add a service area for a partner that already belongs to the neighbourhood is a redundant.
> 
> It does prevent a `neighbourhood_admin` from adding lots of partners to someone else's neighbourhood that they have not been given admin privileges for so I suppose it is an improvement for the rival `neighbourhood_admin`
> 
> probably out of scope from this ticket to make any more changes to what a `neighbourhood_admin` can do to a partner

I guess the use case for this is, an admin manages a larger area, like manchester, and wants to say that one of the partners in their area covers the whole area or some other part of manchester. If the partner covers areas outside of manchester in addition to this they can add them themselves

**Pull request merged**